barrels/Semicircles/Semicircles.mc

```diff
@@ -0,0 +1,257 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+using Toybox.Position;
+using Toybox.Lang;
+using Toybox.Math;
+using Toybox.System;
+using Toybox.Test;
+
+module Semicircles {
+
+    //! Execption for unhandled coordinate formats
+    class BadCoordinateFormatException extends Lang.Exception {
+        private var _msg;
+        public function initialize(msg) {
+            Exception.initialize();
+            _msg = msg;
+        }
+
+        public function getErrorMessage() {
+            return _msg;
+        }
+    }
+
+
+    //! Class to represent a coordinate that does not use Position module
+    //! This class uses 32 bit semicircles, where PI radians are equal
+    //! to 0x80000000 32 bit semicircles. Semicircles handle wrap-around via
+    //! integer overflow, and all angle math can be done using integer math
+    //! speeding up computation. This class also includes a fast distance
+    //! between two point computation.
+    class Coordinate {
+        //! Constant for semicircles per meter at the equator
+        const SC_PER_M = 107.26d;
+
+        private var _lat;
+        private var _lon;
+
+        //! Private conversion function to convert locations to semicircles
+        //! @param [Toybox::Lang::Object] Type based off the following options
+        //!    [Toybox::Lang::Dictionary] Allows input options
+        //!      [:latitude] Specify the latitude
+        //!      [:longitude] Specify the longitude
+        //!      [:format] Either :semicircles or :radians. I spit on your degrees.
+        //!    [Toybox::Position::Location] Convert the location to semicircles
+        //!    [Toybox::Lang::Array] Assumes latitude is in radians as value 0, and longitude is in radians in value 1
+        public function initialize(value) {
+            var coord = convertToSemicircle(value);
+            _lat = coord[0];
+            _lon = coord[1];
+        }
+
+        //! Private conversion function to convert locations to semicircles
+        //! @param [Toybox::Lang::Object] Type based off the following options
+        //!    [Toybox::Position::Location] Convert the location to semicircles
+        //!    [Toybox::Lang::Array] Assumes latitude is in radians as value 0, and longitude is in radians in value 1
+        //!    [Toybox::Lang::Dictionary] Allows input options
+        //!      [:latitude] Specify the latitude
+        //!      [:longitude] Specify the longitude
+        //!      [:format] Either :semicircles or :radians. I spit on your degrees.
+        private function convertToSemicircle(value) {
+            var result = new [2];
+            // Switch based on the object type
+
+            switch(value) {
```

Comment:
A switch statement is probably the right choice from a design perspective but would it make sense to use a chain of if/else statements given they use less memory and an end user may need library code to be as optimized as possible?

Reply:
Why would the switch statement take more memory or code space?  The code expands to the same as a if/elseif block, and even optimizes for switching on instanceof.

Reply:
In addition to the forum thread Brandon linked to there's this thread: https://forums.garmin.com/forum/developers/connect-iq/1232387-switch-vs-if

I don't have proof that they are using more (other than what's in the thread), I don't think they should be using more and in this case maybe they aren't. If they are using more memory in the case it may be close enough to if/else statements that it's not worth taking the hit in code readability. However, if they are using more then it seemed like a question that should be asked.

Reply:
We should not be discouraging people to use switch statements - they are more powerful and provide optimizations we can't do with if statements.